### PR TITLE
Implement initial-resize-group-by feature

### PR DESCRIPTION
**Pull request by bells17**
- https://github.com/topolvm/pvc-autoresizer/pull/174

### CODE REVIEW

File: hooks/persistentvolumeclaim.go
Comment (peng225) [inline hooks/persistentvolumeclaim.go:42]:
```suggestion
	pvcList := &corev1.PersistentVolumeClaimList{}
```

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go:90]: cf. https://github.com/topolvm/pvc-autoresizer/pull/176#discussion_r1118373861

**Review by cupnes — COMMENTED** [review]
The ["Run chart-testing (install)"](https://github.com/topolvm/pvc-autoresizer/blob/f158a7dda4eb3912b5108b952352c68088e866f7/.github/workflows/helm.yaml#L50-L51) CI failed because `ct` simply uses the latest chart version image that has been released. Perhaps this CI should be executed at the time of the chart release.

File: charts/pvc-autoresizer/Chart.yaml
Comment (bells17) [reply charts/pvc-autoresizer/Chart.yaml:32]: I've applied this suggestion.

File: charts/pvc-autoresizer/templates/controller/service.yaml
Comment (bells17) [reply charts/pvc-autoresizer/templates/controller/service.yaml]: I've applied this suggestion.

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I've applied this suggestion.

https://github.com/topolvm/pvc-autoresizer/pull/176#discussion_r1118376222

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: https://github.com/topolvm/pvc-autoresizer/pull/176#discussion_r1118376222

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: https://github.com/topolvm/pvc-autoresizer/pull/176#discussion_r1118376222

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I've applied this suggestion.

File: charts/pvc-autoresizer/values.yaml
Comment (bells17) [reply charts/pvc-autoresizer/values.yaml]: I've applied this suggestion.

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I've applied this suggestion.

File: charts/pvc-autoresizer/templates/controller/deployment.yaml
Comment (bells17) [reply charts/pvc-autoresizer/templates/controller/deployment.yaml]: I've applied this suggestion.

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I modified the domain name and the webhook path.

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I think we need to list the PVC by the `resize.topolvm.io/initial-resize-group-by` annotation from reading this proposal.
So I don't think `client.ListOptions.LabelSelector` can be used, is there something I'm missing in my understanding?

```yaml
kind: PersistentVolumeClaim
metadata:
  name: pvc-x-1
  labels:
    label-foobar: group-x
  annotations:
    resize.topolvm.io/initial-resize-group-by: label-foobar
spec:
  resources:
    requests:
      storage: 20Gi
```

Comment (peng225) [reply hooks/persistentvolumeclaim.go]: In my understanding, the value of `resize.topolvm.io/initial-resize-group-by` annotation is the key of the label which is used for grouping. The grouping itself is done by the label's value.

In the example in the proposal document, the value of `resize.topolvm.io/initial-resize-group-by` is `label-foobar`, so the key of the label for grouping is `label-foobar`.

For example, `pvc-x-1` and `pvc-x-2` have the same label value `group-x`, thus they belong to the same group.

File: e2e/suite_test.go
Comment (peng225) [reply e2e/suite_test.go:356]: OK, thanks.
So, you need not to append the result of `createPodPVC` at the following line?
https://github.com/topolvm/pvc-autoresizer/pull/176/files#diff-df24cc5d6cd0a9d1708368dc0c4119fc7ac5ef7ef7293bc004d64985d736739fR346

File: charts/pvc-autoresizer/README.md
Comment (peng225) [reply charts/pvc-autoresizer/README.md]: Sorry, but I think the option should be remained. I wanted to change the name to something like `pvcMutatingWebhook`.

File: e2e/suite_test.go
Comment (bells17) [reply e2e/suite_test.go:356]: Yes. I need to fix this line.

File: e2e/suite_test.go
Comment (bells17) [reply e2e/suite_test.go:359]: It seems this check already expects 3GiB.

File: hooks/persistentvolumeclaim.go
Comment (bells17) [reply hooks/persistentvolumeclaim.go]: I see. Thank you.

File: charts/pvc-autoresizer/README.md
Comment (peng225) [reply charts/pvc-autoresizer/README.md]: fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: fixed.

File: e2e/suite_test.go
Comment (peng225) [reply e2e/suite_test.go]: Changed to use checkDiskResize.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go:93]: I deleted it.

File: e2e/suite_test.go
Comment (peng225) [reply e2e/suite_test.go:356]: I fixed it.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: I deleted it.

**Comment by peng225** [comment]
@cupnes Thank you for your investigation. I let `ct install` test work only when the source branch name starts with `bump-chart-`.

File: runners/constants.go
Comment (peng225) [inline runners/constants.go:7]: This constant is not used.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [inline hooks/persistentvolumeclaim.go:62]: This comment can be removed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [inline hooks/persistentvolumeclaim.go:66]:
```suggestion
	// Check if the resized capacity is less than the storage limit
```

File: e2e/suite_test.go
Comment (toshipp) [inline e2e/suite_test.go:443]: Why not use bool value directly?

File: e2e/testdata/pod-pvc-template.yaml
Comment (toshipp) [inline e2e/testdata/pod-pvc-template.yaml:6]: if we change type of `useLabel` to bool, we can write it simply.
```suggestion
{{ if eq .useLabel }}
```

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:96]: GetClient returns caching client, but webhook logic needs to read fresh data, so use GetAPIReader instead, IMO.

Comment (toshipp) [reply hooks/persistentvolumeclaim.go:96]: APIReader reads api-server directly, so the load of api-server would be high, but api-server indexes objects by labels, I think it is ok.

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:75]: I think it is better to allow the request and use `storageLimit` as a capacity if the requested capacity is larger than or equal to the limit.

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:61]: It is hard to read the code for me...
How about this？

```suggestion
	newSize := requestedSize
	for _, item := range pvcList.Items {
		if item.Spec.Resources.Requests.Storage().Value() > newSize {
			newSize = item.Spec.Resources.Requests[corev1.ResourceStorage]
		}
	}
	if newSize == requestedSize {
	...
	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = newSize
```

**Review by toshipp — CHANGES_REQUESTED** [review]

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: fixed.

File: charts/pvc-autoresizer/templates/controller/mutatingwebhookconfiguration.yaml
Comment (peng225) [reply charts/pvc-autoresizer/templates/controller/mutatingwebhookconfiguration.yaml]: Thanks. I fixed it.

File: e2e/suite_test.go
Comment (peng225) [reply e2e/suite_test.go]: I fixed it.
(I was imitating other code in this function without deep thoughts...)

File: e2e/testdata/pod-pvc-template.yaml
Comment (peng225) [reply e2e/testdata/pod-pvc-template.yaml]: Fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Thank you for your idea.
I fixed the code in almost the same way as you suggested but used `resourece.Quantity` type of variables instead.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go:67]: I noticed that I misunderstood the spec.. I reverted it and fixed the relevant code.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Thank you for your suggestion. That sounds good. I fixed it.

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:74]: To reduce the api-server load, it is better to move this check before `m.apiReader.List`

**Review by llamerada-jp — CHANGES_REQUESTED** [review]

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Thanks. I fixed it.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go:74]: Thanks. I fixed it.

File: README.md
Comment (peng225) [reply README.md:191]: I added an explanation to README.md.

File: e2e/suite_test.go
Comment (peng225) [inline e2e/suite_test.go:381]: I should add a test case to check that the resize is limited by the storage limit annotation.

**Comment by peng225** [comment]
I will squash the commits after I get approve from all of reviewers.
Please do not merge it until that.

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:64]: Can we write it as `pvc.Spec.Resources.Requests.Storage()`?

File: hooks/persistentvolumeclaim.go
Comment (toshipp) [inline hooks/persistentvolumeclaim.go:68]: ditto

Comment (toshipp) [inline hooks/persistentvolumeclaim.go:68]: It can be simplified as:
```suggestion
		if itemSize := item.Spec.Resources.Requests.Storage(); itemSize.Cmp(newSize) > 0 {
			newSize = itemSize
```

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Fixed.

File: hooks/persistentvolumeclaim.go
Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Fixed.

Comment (peng225) [reply hooks/persistentvolumeclaim.go]: Fixed.

**Review by toshipp — APPROVED** [review]

**Review by llamerada-jp — APPROVED** [review]

**Pull request merged**
